File: packages/plugin-agent/src/server/tools/runPlans.ts
import type { AgentToolDefinition, AgentToolHandler, AgentToolContext, AgentToolResult } from '@purplesector/plugin-api';

export const runPlanToolDefinitions: AgentToolDefinition[] = [
  {
    name: 'listRunPlans',
    description: 'List all run plans for the current user.',
    inputSchema: {
      type: 'object',
      properties: {
        eventId: { type: 'string', description: 'Optional event ID to filter by' },
      },
    },
    category: 'runPlans',
    mutating: false,
  },
  {
    name: 'getRunPlan',
    description: 'Get a run plan with all its items (session/vehicle-config combinations).',
    inputSchema: {
      type: 'object',
      properties: {
        runPlanId: { type: 'string', description: 'The run plan ID' },
      },
      required: ['runPlanId'],
    },
    category: 'runPlans',
    mutating: false,
  },
  {
    name: 'createRunPlan',
    description: 'Create a new run plan — a named collection of planned session/vehicle-configuration combinations.',
    inputSchema: {
      type: 'object',
      properties: {
        name: { type: 'string', description: 'Run plan name' },
        description: { type: 'string' },
        eventId: { type: 'string', description: 'Optional event ID to associate with' },
        items: {
          type: 'array',
          description: 'Array of planned items',
          items: {
            type: 'object',
            properties: {
              sessionName: { type: 'string', description: 'Planned session name' },
              vehicleId: { type: 'string' },
              vehicleConfigurationId: { type: 'string' },
              vehicleSetupId: { type: 'string' },
              notes: { type: 'string' },
            },
            required: ['sessionName'],
          },
        },
      },
      required: ['name'],
    },
    category: 'runPlans',
    mutating: true,
  },
  {
    name: 'updateRunPlan',
    description: 'Update a run plan (name, description, status) or replace its items.',
    inputSchema: {
      type: 'object',
      properties: {
        runPlanId: { type: 'string', description: 'The run plan ID' },
        name: { type: 'string' },
        description: { type: 'string' },
        status: { type: 'string', description: '"draft" | "active" | "completed"' },
        items: {
          type: 'array',
          description: 'If provided, replaces all items',
          items: {
            type: 'object',
            properties: {
              sessionName: { type: 'string' },
              vehicleId: { type: 'string' },
              vehicleConfigurationId: { type: 'string' },
              vehicleSetupId: { type: 'string' },
              notes: { type: 'string' },
            },
            required: ['sessionName'],
          },
        },
      },
      required: ['runPlanId'],
    },
    category: 'runPlans',
    mutating: true,
  },
  {
    name: 'deleteRunPlan',
    description: 'Delete a run plan and all its items.',
    inputSchema: {
      type: 'object',
      properties: {
        runPlanId: { type: 'string', description: 'The run plan ID to delete' },
      },
      required: ['runPlanId'],
    },
    category: 'runPlans',
    mutating: true,
  },
];

export function createRunPlanToolHandlers(prisma: any): Record<string, AgentToolHandler> {
  return {
    async listRunPlans(args: Record<string, unknown>, ctx: AgentToolContext): Promise<AgentToolResult> {
      const where: Record<string, unknown> = { userId: ctx.userId };
      if (args.eventId) where.eventId = args.eventId;

      const plans = await prisma.runPlan.findMany({
        where,
        include: { _count: { select: { items: true } }, event: { select: { id: true, name: true } } },
        orderBy: { createdAt: 'desc' },
      });
      return { success: true, data: plans, message: `Found ${plans.length} run plans.` };
    },

    async getRunPlan(args: Record<string, unknown>, ctx: AgentToolContext): Promise<AgentToolResult> {
      const plan = await prisma.runPlan.findFirst({
        where: { id: args.runPlanId as string, userId: ctx.userId },
        include: {
          event: { select: { id: true, name: true } },
          items: {
            orderBy: { order: 'asc' },
            include: {
              vehicle: { select: { id: true, name: true } },
              vehicleConfiguration: { select: { id: true, name: true } },
              vehicleSetup: { select: { id: true, name: true } },
              session: { select: { id: true, name: true, status: true } },
            },
          },
        },
      });
      if (!plan) return { success: false, message: 'Run plan not found.' };
      return { success: true, data: plan };
    },

    async createRunPlan(args: Record<string, unknown>, ctx: AgentToolContext): Promise<AgentToolResult> {
      const items = (args.items as Array<Record<string, unknown>>) || [];

      const plan = await prisma.runPlan.create({
        data: {
          userId: ctx.userId,
          name: args.name as string,
          description: (args.description as string) || null,
          eventId: (args.eventId as string) || null,
          status: 'draft',
          items: {
            create: items.map((item, index) => ({
              order: index,
              sessionName: item.sessionName as string,
              vehicleId: (item.vehicleId as string) || null,
              vehicleConfigurationId: (item.vehicleConfigurationId as string) || null,
              vehicleSetupId: (item.vehicleSetupId as string) || null,
              notes: (item.notes as string) || null,
            })),
          },
        },
        include: { items: { orderBy: { order: 'asc' } } },
      });
      return { success: true, data: plan, message: `Created run plan "${plan.name}" with ${plan.items.length} items.` };
    },

    async updateRunPlan(args: Record<string, unknown>, ctx: AgentToolContext): Promise<AgentToolResult> {
      const { runPlanId, items, ...updates } = args;
      const existing = await prisma.runPlan.findFirst({ where: { id: runPlanId as string, userId: ctx.userId } });
      if (!existing) return { success: false, message: 'Run plan not found.' };

      const data: Record<string, unknown> = {};
      if (updates.name !== undefined) data.name = updates.name;
      if (updates.description !== undefined) data.description = updates.description;
      if (updates.status !== undefined) data.status = updates.status;

      if (items !== undefined) {
        await prisma.runPlanItem.deleteMany({ where: { runPlanId: runPlanId as string } });
        const itemsArr = items as Array<Record<string, unknown>>;
        await prisma.runPlanItem.createMany({
          data: itemsArr.map((item, index) => ({
            runPlanId: runPlanId as string,
            order: index,
            sessionName: item.sessionName as string,
            vehicleId: (item.vehicleId as string) || null,
            vehicleConfigurationId: (item.vehicleConfigurationId as string) || null,
            vehicleSetupId: (item.vehicleSetupId as string) || null,
            notes: (item.notes as string) || null,
          })),
        });
      }

      const plan = await prisma.runPlan.update({
        where: { id: runPlanId as string },
        data,
        include: { items: { orderBy: { order: 'asc' } } },
      });
      return { success: true, data: plan, message: `Updated run plan "${plan.name}".` };
    },

    async deleteRunPlan(args: Record<string, unknown>, ctx: AgentToolContext): Promise<AgentToolResult> {
      const existing = await prisma.runPlan.findFirst({ where: { id: args.runPlanId as string, userId: ctx.userId } });
      if (!existing) return { success: false, message: 'Run plan not found.' };
      await prisma.runPlan.delete({ where: { id: args.runPlanId as string } });
      return { success: true, message: `Deleted run plan "${existing.name}".` };
    },
  };
}

File: apps/web/apps/agent/tools/run_plans.py
"""Agent tools for Run Plan CRUD."""

from __future__ import annotations

from apps.events.models import RunPlan, RunPlanItem

from . import ToolDefinition, ToolContext, ToolResult, register_tools

definitions = [
    ToolDefinition(
        name="listRunPlans",
        description="List all run plans for the current user.",
        input_schema={
            "type": "object",
            "properties": {
                "eventId": {"type": "string", "description": "Optional event ID to filter by"},
            },
        },
        category="runPlans",
    ),
    ToolDefinition(
        name="getRunPlan",
        description="Get a run plan with all its items (session/vehicle-config combinations).",
        input_schema={
            "type": "object",
            "properties": {
                "runPlanId": {"type": "string", "description": "The run plan ID"},
            },
            "required": ["runPlanId"],
        },
        category="runPlans",
    ),
    ToolDefinition(
        name="createRunPlan",
        description="Create a new run plan — a named collection of planned session/vehicle-configuration combinations.",
        input_schema={
            "type": "object",
            "properties": {
                "name": {"type": "string", "description": "Run plan name"},
                "description": {"type": "string"},
                "eventId": {"type": "string", "description": "Optional event ID to associate with"},
                "items": {
                    "type": "array",
                    "description": "Array of planned items",
                    "items": {
                        "type": "object",
                        "properties": {
                            "sessionName": {"type": "string", "description": "Planned session name"},
                            "vehicleId": {"type": "string"},
                            "vehicleConfigurationId": {"type": "string"},
                            "vehicleSetupId": {"type": "string"},
                            "notes": {"type": "string"},
                        },
                        "required": ["sessionName"],
                    },
                },
            },
            "required": ["name"],
        },
        category="runPlans",
        mutating=True,
    ),
    ToolDefinition(
        name="updateRunPlan",
        description="Update a run plan (name, description, status) or replace its items.",
        input_schema={
            "type": "object",
            "properties": {
                "runPlanId": {"type": "string", "description": "The run plan ID"},
                "name": {"type": "string"},
                "description": {"type": "string"},
                "status": {"type": "string", "description": '"draft" | "active" | "completed"'},
                "items": {
                    "type": "array",
                    "description": "If provided, replaces all items",
                    "items": {
                        "type": "object",
                        "properties": {
                            "sessionName": {"type": "string"},
                            "vehicleId": {"type": "string"},
                            "vehicleConfigurationId": {"type": "string"},
                            "vehicleSetupId": {"type": "string"},
                            "notes": {"type": "string"},
                        },
                        "required": ["sessionName"],
                    },
                },
            },
            "required": ["runPlanId"],
        },
        category="runPlans",
        mutating=True,
    ),
    ToolDefinition(
        name="deleteRunPlan",
        description="Delete a run plan and all its items.",
        input_schema={
            "type": "object",
            "properties": {
                "runPlanId": {"type": "string", "description": "The run plan ID to delete"},
            },
            "required": ["runPlanId"],
        },
        category="runPlans",
        mutating=True,
    ),
]


def _serialize_plan(plan: RunPlan, include_items: bool = False) -> dict:
    d = {
        "id": str(plan.id),
        "name": plan.name,
        "description": plan.description,
        "status": plan.status,
        "eventId": str(plan.event_id) if plan.event_id else None,
        "createdAt": plan.created_at.isoformat(),
    }
    if include_items:
        items = plan.items.order_by("order").select_related("vehicle", "vehicle_configuration", "vehicle_setup", "session")
        d["items"] = [
            {
                "id": str(item.id),
                "order": item.order,
                "sessionName": item.session_name,
                "vehicleId": str(item.vehicle_id) if item.vehicle_id else None,
                "vehicle": {"id": str(item.vehicle.id), "name": item.vehicle.name} if item.vehicle else None,
                "vehicleConfigurationId": str(item.vehicle_configuration_id) if item.vehicle_configuration_id else None,
                "vehicleConfiguration": {"id": str(item.vehicle_configuration.id), "name": item.vehicle_configuration.name} if item.vehicle_configuration else None,
                "vehicleSetupId": str(item.vehicle_setup_id) if item.vehicle_setup_id else None,
                "vehicleSetup": {"id": str(item.vehicle_setup.id), "name": item.vehicle_setup.name} if item.vehicle_setup else None,
                "notes": item.notes,
                "session": {"id": str(item.session.id), "name": item.session.name, "status": item.session.status} if item.session else None,
            }
            for item in items
        ]
    return d


async def list_run_plans(args: dict, ctx: ToolContext) -> ToolResult:
    from django.db.models import Count

    qs = RunPlan.objects.filter(user_id=ctx.user_id).annotate(item_count=Count("items")).select_related("event")
    if args.get("eventId"):
        qs = qs.filter(event_id=args["eventId"])
    plans = qs.order_by("-created_at")
    data = [
        {
            **_serialize_plan(p),
            "event": {"id": str(p.event.id), "name": p.event.name} if p.event else None,
            "_count": {"items": p.item_count},
        }
        for p in plans
    ]
    return ToolResult(success=True, data=data, message=f"Found {len(data)} run plans.")


async def get_run_plan(args: dict, ctx: ToolContext) -> ToolResult:
    try:
        plan = RunPlan.objects.select_related("event").get(id=args["runPlanId"], user_id=ctx.user_id)
    except RunPlan.DoesNotExist:
        return ToolResult(success=False, message="Run plan not found.")
    d = _serialize_plan(plan, include_items=True)
    d["event"] = {"id": str(plan.event.id), "name": plan.event.name} if plan.event else None
    return ToolResult(success=True, data=d)


async def create_run_plan(args: dict, ctx: ToolContext) -> ToolResult:
    items_data = args.get("items") or []
    plan = RunPlan.objects.create(
        user_id=ctx.user_id,
        name=args["name"],
        description=args.get("description") or None,
        event_id=args.get("eventId") or None,
        status="draft",
    )
    for i, item in enumerate(items_data):
        RunPlanItem.objects.create(
            run_plan=plan,
            order=i,
            session_name=item["sessionName"],
            vehicle_id=item.get("vehicleId") or None,
            vehicle_configuration_id=item.get("vehicleConfigurationId") or None,
            vehicle_setup_id=item.get("vehicleSetupId") or None,
            notes=item.get("notes") or None,
        )
    return ToolResult(
        success=True,
        data=_serialize_plan(plan, include_items=True),
        message=f'Created run plan "{plan.name}" with {len(items_data)} items.',
    )


async def update_run_plan(args: dict, ctx: ToolContext) -> ToolResult:
    try:
        plan = RunPlan.objects.get(id=args["runPlanId"], user_id=ctx.user_id)
    except RunPlan.DoesNotExist:
        return ToolResult(success=False, message="Run plan not found.")

    if "name" in args:
        plan.name = args["name"]
    if "description" in args:
        plan.description = args["description"]
    if "status" in args:
        plan.status = args["status"]
    plan.save()

    if "items" in args:
        plan.items.all().delete()
        for i, item in enumerate(args["items"]):
            RunPlanItem.objects.create(
                run_plan=plan,
                order=i,
                session_name=item["sessionName"],
                vehicle_id=item.get("vehicleId") or None,
                vehicle_configuration_id=item.get("vehicleConfigurationId") or None,
                vehicle_setup_id=item.get("vehicleSetupId") or None,
                notes=item.get("notes") or None,
            )

    return ToolResult(success=True, data=_serialize_plan(plan, include_items=True), message=f'Updated run plan "{plan.name}".')


async def delete_run_plan(args: dict, ctx: ToolContext) -> ToolResult:
    try:
        plan = RunPlan.objects.get(id=args["runPlanId"], user_id=ctx.user_id)
    except RunPlan.DoesNotExist:
        return ToolResult(success=False, message="Run plan not found.")
    name = plan.name
    plan.delete()
    return ToolResult(success=True, message=f'Deleted run plan "{name}".')


handlers = {
    "listRunPlans": list_run_plans,
    "getRunPlan": get_run_plan,
    "createRunPlan": create_run_plan,
    "updateRunPlan": update_run_plan,
    "deleteRunPlan": delete_run_plan,
}

register_tools(definitions, handlers)
